### PR TITLE
Fix GenericType resolution in parameterized classes

### LINKED ISSUE #2305: GenericType don't work if they are created in a parameterized class
Things that make one cry:

```java
 @Test
    void thisWorks() {

        class Foo {
            Type getType() {
                return new GenericType<List<String>>() {}.getType();
            }
        }

        assertThat (new Foo().getType()).isNotNull();
    }

    @Test
    void thisDoesNot() {

        class Foo<T> {
            Type getType() {
                return new GenericType<List<String>>() {}.getType();
            }
        }

        assertThat (new Foo<>().getType()).isNotNull();
    }
```

This seems to be a bug in geantryref.

### PR DESCRIPTION
**Pull request by hgschmie**
Port the code from geantryref.TypeToken#extractType for classes passed into GenericType.

Fixes #2305

### CODE REVIEW

**Comment by hgschmie** [comment]
This is a straight port of the code in TypeToken#extractType in geantryref. If the type is not a naked class, it falls back to the existing code.

File: core/src/test/java/org/jdbi/v3/core/generic/GenericTypesTest.java [hunk 1]
```diff
@@ -77,6 +78,17 @@ public void findMultipleGenericParameters() throws NoSuchMethodException {
...
 
+    @Test
+    void testGenericTypeWorksInParameterizedClasses() {
+        class Foo<T> {
+            Type getType() {
+                return new GenericType<List<String>>() {}.getType();
+            }
+        }
+
+        assertThat(new Foo<>().getType()).isNotNull();
```
[the remark this answers is not in the document]
Comment (hgschmie) [reply core/src/test/java/org/jdbi/v3/core/generic/GenericTypesTest.java:89]: sure. This was just the "works / explodes" check. :-)

**Comment by kaqqao** [comment]
Hey there 👋
As I commented on the related GeantyRef issue, what you're seeing is actually the expected behavior. But it sounds like you could use a new method that tolerates partially resolved types. If that's correct, I can add it to GeantyRef.

As a side note, in your current code, since you're dealing with `Type`s only, you don't have to bother with `getAnnotatedSuperclass`. It's enough to do:

```java
ParameterizedType superType = (ParameterizedType) getClass().getGenericSuperclass();
return superType.getActualTypeArguments()[0];
```

**Comment by stevenschlansker** [comment]
Thank you for the update! I simplified the code a bit in #2321
If you think other users would also enjoy this functionality and would like to add such a method we'd love to use it, but can live with the situation as it is.

**Pull request merged**